File: sdk/src/main/java/co/omise/android/models/Token.java
package co.omise.android.models;

import android.os.Parcel;
import android.os.Parcelable;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.json.JSONException;
import org.json.JSONObject;

import co.omise.android.SDKLog;
import co.omise.android.api.RequestBuilder;
import okhttp3.FormBody;
import okhttp3.HttpUrl;
import okhttp3.RequestBody;

public class Token extends Model {
    public final boolean used;
    public final Card card;

    public Token(String rawJson) throws JSONException {
        this(new JSONObject(rawJson));
    }

    private Token(JSONObject json) throws JSONException {
        super(json);
        card = json.has("card") ? new Card(json.getJSONObject("card")) : null;
        used = JSON.bool(json, "used");
    }

    public static class CreateTokenRequestBuilder extends RequestBuilder<Token> {
        String name;
        String number;
        int expirationMonth;
        int expirationYear;
        String securityCode;
        String city;
        String postalCode;

        public CreateTokenRequestBuilder(String name, String number, int expirationMonth, int expirationYear, String securityCode) {
            this.name = name;
            this.number = number;
            this.expirationMonth = expirationMonth;
            this.expirationYear = expirationYear;
            this.securityCode = securityCode;
        }

        @NotNull
        @Override
        protected HttpUrl path() {
            return HttpUrl.parse("https://vault.omise.co/tokens");
        }

        @Nullable
        @Override
        public RequestBody payload() {
            FormBody.Builder builder = new FormBody.Builder()
                    .add("card[name]", name)
                    .add("card[number]", number)
                    .add("card[expiration_month]", Integer.toString(expirationMonth))
                    .add("card[expiration_year]", Integer.toString(expirationYear))
                    .add("card[security_code]", securityCode);

            if (city != null && !city.isEmpty()) {
                builder.add("card[city]", city);
            }
            if (postalCode != null && !postalCode.isEmpty()) {
                builder.add("card[postal_code]", postalCode);
            }

            return builder.build();
        }

        @NotNull
        @Override
        public String method() {
            return RequestBuilder.POST;
        }

        @NotNull
        @Override
        protected Class<Token> type() {
            return Token.class;
        }

        public CreateTokenRequestBuilder city(String city) {
            this.city = city;
            return this;
        }

        public CreateTokenRequestBuilder postalCode(String postalCode) {
            this.postalCode = postalCode;
            return this;
        }
    }

    public static final Parcelable.Creator<Token> CREATOR = new Creator<Token>() {
        @Override
        public Token createFromParcel(Parcel source) {
            try {
                return new Token(source.readString());
            } catch (JSONException e) {
                SDKLog.wtf("failed to deparcelize Token object", e);
                return null;
            }
        }

        @Override
        public Token[] newArray(int size) {
            return new Token[size];
        }
    };
}

File: sdk/src/main/java/co/omise/android/models/Token.kt
package co.omise.android.models

import co.omise.android.api.Endpoint
import co.omise.android.api.RequestBuilder
import com.fasterxml.jackson.annotation.JsonProperty
import kotlinx.android.parcel.Parcelize
import kotlinx.android.parcel.TypeParceler
import okhttp3.HttpUrl
import okhttp3.RequestBody
import org.joda.time.DateTime
import java.io.IOException

/**
 * Represents Token object and contains its [RequestBuilder].
 *
 * @see [Token API](https://www.omise.co/tokens-api)
 */
@Parcelize
@TypeParceler<ChargeStatus, ChargeStatusParceler>()
data class Token(
    val used: Boolean = false,
    val card: Card? = null,
    @field:JsonProperty("charge_status")
    val chargeStatus: ChargeStatus = ChargeStatus.Unknown,
    override var modelObject: String? = null,
    override var id: String? = null,
    override var livemode: Boolean = false,
    override var location: String? = null,
    override var created: DateTime? = null,
    override var deleted: Boolean = false,
) : Model {
    /**
     * The [RequestBuilder] class for creating a Token.
     */
    class CreateTokenRequestBuilder(
        @field:JsonProperty("card")
        val card: CardParam? = null,
        @field:JsonProperty("tokenization")
        val tokenization: TokenizationParam? = null,
    ) : RequestBuilder<Token>() {
        override fun path(): HttpUrl {
            return buildUrl(Endpoint.VAULT, "tokens")
        }

        @Throws(IOException::class)
        override fun payload(): RequestBody? {
            return serialize()
        }

        override fun method(): String {
            return POST
        }

        override fun type(): Class<Token> {
            return Token::class.java
        }
    }

    /**
     * The [RequestBuilder] class for retrieving a particular Token.
     * @property id the Token ID.
     */
    class GetTokenRequestBuilder(val id: String) : RequestBuilder<Token>() {
        override fun path(): HttpUrl {
            return buildUrl(Endpoint.VAULT, "tokens", id)
        }

        override fun method(): String {
            return GET
        }

        override fun type(): Class<Token> {
            return Token::class.java
        }
    }
}
